File: physics_engine.py
# physics_engine.py

class PhysicsEngine:
    def __init__(self):
        self.gravity = 9.81
        self.objects = []

    def add_object(self, obj):
        self.objects.append(obj)

    def apply_gravity(self):
        for obj in self.objects:
            if obj.mass:
                obj.apply_force((0, obj.mass * self.gravity))

class GameObject:
    def __init__(self, mass=0, position=(0, 0)):
        self.mass = mass
        self.position = position
        self.velocity = (0, 0)

    def apply_force(self, force):
        acceleration = (force[0] / self.mass, force[1] / self.mass)
        self.velocity = (self.velocity[0] + acceleration[0], self.velocity[1] + acceleration[1])

    def update(self, time_step):
        new_x = self.position[0] + self.velocity[0] * time_step
        new_y = self.position[1] + self.velocity[1] * time_step
        self.position = (new_x, new_y)

    def __str__(self):
        return f"Object with mass {self.mass} at position {self.position}"

if __name__ == "__main__":
    engine = PhysicsEngine()
    obj1 = GameObject(mass=5, position=(0, 0))
    obj2 = GameObject(mass=10, position=(3, 0))

    engine.add_object(obj1)
    engine.add_object(obj2)

    for _ in range(10):
        engine.apply_gravity()
        obj1.update(0.1)
        obj2.update(0.1)
        print(obj1)
        print(obj2)

File: physics_engine.js
// Директория: PiroEngine/physics_engine.js

class PhysicsEngine {
  constructor() {
    this.objects = [];
    this.gravity = 9.81; // Гравитационная постоянная (м/с^2)
  }

  // Функция для добавления объектов в симуляцию
  addObject(object) {
    this.objects.push(object);
  }

  // Функция для обновления физической модели объектов
  update() {
    for (const object of this.objects) {
      // Применяем гравитацию
      object.velocity.y -= this.gravity;

      // Здесь можно добавить другие физические эффекты, такие как столкновения и динамику движения
    }
  }
}

class GameObject {
  constructor() {
    this.position = { x: 0, y: 0, z: 0 };
    this.velocity = { x: 0, y: 0, z: 0 };
    this.mass = 1; // Масса объекта (кг)
  }

  // Функция для обновления положения объекта
  updatePosition() {
    this.position.x += this.velocity.x;
    this.position.y += this.velocity.y;
    this.position.z += this.velocity.z;
  }
}

// Пример использования PhysicsEngine
const engine = new PhysicsEngine();

const object1 = new GameObject();
object1.mass = 2;
engine.addObject(object1);

const object2 = new GameObject();
object2.mass = 1;
engine.addObject(object2);

// Обновление физической модели
engine.update();
object1.updatePosition();
object2.updatePosition();

console.log('Позиция объекта 1:', object1.position);
console.log('Позиция объекта 2:', object2.position);
